### ISSUE TITLE
Weird behavior if selection starts in an element which disallows given attribute

### ISSUE THREAD

**Issue by scofalik** [ckeditor/ckeditor5-basic-styles#56]
Consider following scenario:
```html
<h1>Tit[le</h1>
<p>F]oo bar.</p>
```

Selection starts in heading and ends in paragraph. Assume that `bold` is disallowed in heading.

As of now, you are able to bold this selection, but you are unable to unbold it. This is because command's `value` is taken from first position but `isEnabled` from whole selection. So before you bold, command is enabled, so you bold the selection. But, the part in heading is not bold. So the `value` is still `false` (not bolded). Using the command again does not change anything because again and again bold is applied.

As much as we had good reasons for implementing those properties as we did, the behavior seems strange and buggy. We need to change it. We should either scan the whole range when checking `value` or change how `isEnabled` works.

Right now `isEnabled` is `true` if there is at least one place in selection where an attribute can be applied. Maybe we should change so that it is `false` whenever there is at least one place in selection where an attribute cannot be applied. Or make it `false` if selection starts in a place where the attribute cannot be applied.

**Comment by Reinmar** [comment]
This was discussed many times and so far we didn't have better ideas.

The same happens e.g. with a selected image – when you select it, click bold, the caption is bolded, but the value of the command doesn't change because the bold gets selected somewhere deep inside.

The only solution that I can see right now would be this:

* value – checks whether bold is applied to **all** items in the selection **where it's allowed**,
* isEnabled – whether bold is allowed anywhere in the selection.

That might work. But that's also pretty heavy and it will change behaviour of this:

```html
<p>foo<b>x[x</b>y]y</p>
```

For now, you get the bold on. After such a change it would be off. That's a pretty big change.

**Comment by scofalik** [comment]
There is one more way to fix this one certain scenario but I am afraid that it might leave some other cases uncovered or lead to other strange behavior.

The fix is easy - after you apply the command, "manually" change `this.value`. This will always enable toggling. Since, `this.value` is refreshed automatically on `changesDone`, we should not be stuck with incorrect command's `value`.

**Comment by fredck** [comment]
@Reinmar, your proposal may be counterintuitive in the originally reported case, because the Title would not have bold, but still the bold button would be ON.

**Comment by Reinmar** [comment]
> @Reinmar, your proposal may be counterintuitive in the originally reported case, because the Title would not have bold, but still the bold button would be ON.

I know – pretty much the same with an image where only the caption is bolded. That's why we decided to keep it off.

**Comment by Reinmar** [comment]
> The fix is easy - after you apply the command, "manually" change this.value. This will always enable toggling. Since, this.value is refreshed automatically on changesDone, we should not be stuck with incorrect command's value.

Interesting. It might solve the issue for a second (you see the state of the button to change after you applied bold), but it would be confusing.

You apply bold to a selection, it gets applied to part of it, the command is turned on forcefully, but you move the selection and the command is again off. 

Or you don't move the command but an external change comes (collab editing case) and `changesDone` is fired and your command gets refreshed. So, the command will be on for a short moment only.

**Comment by scofalik** [comment]
Yes that's why I wasn't happy with my proposal.

**Comment by fredck** [comment]
It's almost like we don't have much to do here.

**Comment by Reinmar** [comment]
I think that the issue is valid. But we don't see a viable solution for it. So we can close it, but if anyone will have an idea what we could do – please comment.

**Comment by pjasiun** [comment]
**TL;DR: I think that the `value` should not be defined based on the first element in the range, but the first which allows this style.**

I agree that the perfect solution here does not exist since the range can contain both bolded and not bolded elements, and we will never know if the content should be bolded or unbolded in such situation. However, what is wrong, for sure, is making the decision based on the content which can not be bolded at all. We should scan the range and find the first element which can apply this style. Based on it we should set the current `value`. If there is no content which can be bolded in the range, the button should be disabled.

**Comment by Reinmar** [comment]
I haven't analysed PJ's proposal but to not forget it I'll reopen it.

**Comment by Reinmar** [comment]
DUP reported in https://github.com/ckeditor/ckeditor5/issues/1073.
